### DISCUSSION TITLE
FT_Get_First_Char: symbol not found on alpine

### DISCUSSION THREAD

**Discussion by John-James-Rambo** [pymupdf/PyMuPDF#1071] [Looking for help]
I have install `pymudef` on alpine container. I know there was already an issue https://github.com/pymupdf/PyMuPDF/issues/391 about installing it. I met similar problems but in the end I successfully installed it in the alphine container.

The problem occurred during import of the `fitz`.

## Describe the bug (mandatory)
During import `fitz` I get:
``` python
>>> import fitz
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.7/site-packages/fitz/__init__.py", line 2, in <module>
    from fitz.fitz import *
  File "/usr/local/lib/python3.7/site-packages/fitz/fitz.py", line 15, in <module>
    from . import _fitz
ImportError: Error relocating /usr/local/lib/python3.7/site-packages/fitz/_fitz.cpython-37m-x86_64-linux-gnu.so: FT_Get_First_Char: symbol not found
```

## To Reproduce (mandatory)
``` python
import fitz
```

## Expected behavior (optional)
Import `fitz` should be flawlessly.

## Your configuration (mandatory)
``` Dockerfile
FROM python:3.7-alpine

RUN apk add --no-cache mupdf mupdf-tools mupdf-dev gcc musl-dev freetype-dev jpeg jbig2dec freetype openssl
RUN ln -s /usr/include/freetype2/ft2build.h /usr/include/ft2build.h \
 && ln -s /usr/include/freetype2/freetype/ /usr/include/freetype
RUN pip install pymupdf
```

Unfortunately, I'm not experienced with the Linux builds, so maybe something is wrong with the linking.

**Comment by John-James-Rambo** [comment]
I've also tried to get tar.gz from PyPi and in `setup.py` append `include_dirs` with `"/usr/include/freetype2"`, so I don't need linking to `freetype` in dockerfile anymore.

I've added also `"freetype"` to `DEFAULTS` in `setup.py`  <s>but it didn't helped</s>. It helped, I've just didn't rebuild it (first I made `python setupy.py build` then `pip install . --user`.  
Next error I met was with:

``` python
>>> import fitz
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/root/.local/lib/python3.7/site-packages/fitz/__init__.py", line 2, in <module>
    from fitz.fitz import *
  File "/root/.local/lib/python3.7/site-packages/fitz/fitz.py", line 15, in <module>
    from . import _fitz
ImportError: Error relocating /usr/lib/libmupdf.so.0: opj_stream_set_user_data: symbol not found
```

So I've tried to add to   `"jpeg"` and `"jbig2dec"` to DEFAULTS but in the end I've got

```
    ...
    fitz/fitz_wrap.c:3402:12: warning: 'span_origin.y' may be used uninitialized in this function [-Wmaybe-uninitialized]
     3402 |     return Py_BuildValue("ff", p.x, p.y);
          |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    fitz/fitz_wrap.c:5870:14: note: 'span_origin.y' was declared here
     5870 |     fz_point span_origin;
          |              ^~~~~~~~~~~
    gcc -shared -Wl,--strip-all build/temp.linux-x86_64-3.8/fitz/fitz_wrap.o -L/usr/local/lib -lmupdf -lmupdf-third -lfreetype -ljpeg -ljbig2dec -o build/lib.linux-x86_64-3.8/fitz/_fitz.cpython-38-x86_64-linux-gnu.so
    /usr/lib/gcc/x86_64-alpine-linux-musl/10.2.1/../../../../x86_64-alpine-linux-musl/bin/ld: cannot find -ljpeg
    /usr/lib/gcc/x86_64-alpine-linux-musl/10.2.1/../../../../x86_64-alpine-linux-musl/bin/ld: cannot find -ljbig2dec
    collect2: error: ld returned 1 exit status
```

**Comment by JorjMcKie** [comment]
Cannot reproduce as I have no alpin Linux.
What happens if you **_exactly_** follow the respective installation steps as given in one the scripts [here](https://github.com/pymupdf/PyMuPDF/tree/master/installation)?

Did you try to install from wheel, and it didn't work?

**Comment by JorjMcKie** [comment]
the error ``Error relocating /usr/l ...`` may point to a missing ``export CFLAGS="-fPIC"``.

**Comment by pajakpawel** [comment]
Hi!
I was working on this and here is what I came up with.

1. Modify setup.py:

- Create library set for Alpine Linux like below:
``` python
ALPINE = DEFAULT + [
    "jbig2dec",
    "jpeg",
    "openjp2",
    "harfbuzz",
]
```
In particular the absence of "harfbuzz" library is causing the FT_Get_First_Char error.
- Add new config to LIBRARIES variable:
``` python
LIBRARIES = {
    "default": DEFAULT,
    "ubuntu": DEFAULT,
    "arch": ARCH_LINUX,
    "manjaro": ARCH_LINUX,
    "artix": ARCH_LINUX,
    "opensuse": OPENSUSE,
    "fedora": FEDORA,
    "alpine": ALPINE,
}
```

- Add "/usr/include/freetype2" path to include_dirs:
``` python
if sys.platform.startswith("linux") or "gnu" in sys.platform:
    module = Extension(
        "fitz._fitz",  # name of the module
        ["fitz/fitz_wrap.c"],  # C source file
        include_dirs=[  # we need the path of the MuPDF headers
            "/usr/include/mupdf",
            "/usr/local/include/mupdf",
            "mupdf/thirdparty/freetype/include",
            "/usr/include/freetype2",
        ],
        libraries=load_libraries(),
    )
```
**It would be really appreciated if this setup.py modification could be merged into PyMuPDF.**

2. Dockerfile:
- Build from alpine and add necessary packages:
``` dockerfile
FROM python:3.7-alpine

RUN apk add --no-cache \
    mupdf-dev \
    gcc \
    musl-dev \
    jbig2dec \
    openjpeg-dev \
    jpeg-dev \
    harfbuzz-dev
```

- Download PyMuPDF sources:
``` dockerfile
RUN mkdir -p pymupdf \
 && cd pymupdf \
 && wget https://files.pythonhosted.org/packages/02/5f/ed082e4202b811ae6ba319dba0549f9d5f0c84ed6b54b7657d01bb63cc91/PyMuPDF-1.18.12.tar.gz \
 && tar -zxvf PyMuPDF-1.18.12.tar.gz
```

- Copy modified setup.py into sources directory:
``` dockerfile
COPY setup.py /pymupdf/PyMuPDF-1.18.12
```

- Create symbolic link to libjbig2dec:
``` dockerfile
RUN ln -s /usr/lib/libjbig2dec.so.0 /usr/lib/libjbig2dec.so
```

- Install PyMuPDF and do some cleaning:
``` dockerfile
RUN cd /pymupdf/PyMuPDF-1.18.12 \
 && python setup.py build \
 && pip install --upgrade pip \
 && pip install . --user \
 && cd / \
 && rm -rf pymupdf
```

As I mentioned it would be really great if PyMuPDF setup.py could be modified to support Alpine Linux. 
Is it possible @JorjMcKie ?

**Comment by JorjMcKie** [comment]
>  Is it possible @JorjMcKie?

Hi - thank you very much for this research and solution indeed!
I will certainly include this in the repo as soon as possible.
